Release/include/cpprest/details/basic_types.h

```diff
@@ -43,20 +48,358 @@ typedef uint64_t size64_t;
+    }
+
+    _DECLSPEC_ALLOCATOR pointer allocate(size_type _Count, const void *) {
+        // allocate array of _Count elements, ignore hint
+        return allocate(_Count);
+    }
+
+    void deallocate(pointer _Ptr, size_type _Count) {
+        // deallocate object at _Ptr
+        UNREFERENCED_PARAMETER(_Count);
+        ::HeapFree(::GetProcessHeap(), NULL, _Ptr);
+    }
+};
+
+template <class _Ty1, class _Ty2>
+inline bool operator == (const allocator<_Ty1>&, const allocator<_Ty2>&) {
+    return true;
+}
+
+template <class _Ty1, class _Ty2>
+inline bool operator != (const allocator<_Ty1>& _Left, const allocator<_Ty2>& _Right) {
+    return !(_Left == _Right);
+}
+
+// TEMPLATE CLASS secure_allocator
+template<class _Ty>
+class secure_allocator : public utility::allocator<_Ty> {
+public:
+    typedef secure_allocator<_Ty> _Alloc;
+    typedef utility::allocator<_Ty> _Mybase;
+
+    template<class _Other>
+    struct rebind {
+        // convert this type to secure_allocator<_Other>
+        typedef secure_allocator<_Other> other;
+    };
+
+    secure_allocator() CPPREST_NOEXCEPT : _Mybase() {
+        // construct default secure_allocator (do nothing)
+    }
+
+    secure_allocator(const _Alloc& _Right) CPPREST_NOEXCEPT : _Mybase(_Right) {
+        // construct by copying (do nothing)
+    }
+
+    secure_allocator(_Alloc&& _Right) CPPREST_NOEXCEPT : _Mybase(std::move(_Right)) {
+        // construct by moving (do nothing)
+    }
+
+    template<class _Other>
+    secure_allocator(const secure_allocator<_Other>& _Right) CPPREST_NOEXCEPT : _Mybase(_Right) {
+        // construct from a related secure_allocator (do nothing)
+    }
+
+    template<class _Other>
+    secure_allocator(secure_allocator<_Other>&& _Right) CPPREST_NOEXCEPT : _Mybase(std::forward<_Other>(_Right)) {
+        // construct from a related secure_allocator (do nothing)
+    }
+
+    _Alloc& operator=(const _Alloc& _Right) {
+        // assign by copying
+        _Mybase::operator=(_Right);
+        return (*this);
+    }
+
+    _Alloc& operator=(_Alloc&& _Right) {
+        // assign by moving
+        _Mybase::operator=(std::move(_Right));
+        return (*this);
+    }
+
+    template<class _Other>
+    _Alloc& operator=(const secure_allocator<_Other>&) {
+        // assign from a related secure_allocator
+        _Mybase::operator=(std::forward<_Other>(_Right));
+        return (*this);
+    }
+
+    _DECLSPEC_ALLOCATOR pointer allocate(size_type _Count) {
+        // allocate array of _Count elements
+        return static_cast<pointer>(::HeapAlloc(::GetProcessHeap(), NULL, _Count * sizeof(_Ty)));
+    }
+
+    _DECLSPEC_ALLOCATOR pointer allocate(size_type _Count, const void *) {
+        // allocate array of _Count elements, ignore hint
+        return allocate(_Count);
+    }
+
+    void deallocate(pointer _Ptr, size_type _Count) {
+        // deallocate object at _Ptr
+        ::SecureZeroMemory(_Ptr, _Count * sizeof(_Ty));
```

Comment:
Note that this will zero allocated memory blocks but will not necessarily zero everything a container touches. For example, node-based containers with container-internal sentinel nodes, or std::string's small string optimization.

In either case, this doesn't seem appropriate to include here...

Reply:
The reasoning I included the secure_allocator was for std::vector when encrypting/decrypting user supplied credentials, where for example on line 109 the buffer may be resized and moved to another memory block, how would one ensure that the previous memory block has been securely zeroed ?

https://github.com/Microsoft/cpprestsdk/blob/f897582260f2942284f9213ab9027486bde21b73/Release/src/utilities/web_utilities.cpp#L92-L115

Reply:
Yikes, you're correct, that's broken. :( For std::vector (and only std::vector) I think an allocator like this can work.

That  said I would prefer to just fix that code to make the vector the correct size at the beginning :)

Reply:
Unfortunately this allocator doesn't make the type Win32Encryption safe, and I don't think any change in the memory allocation strategy can do that (for precisely the reason in this comment, the container body itself is *not* zeroed).

It looks like all the APIs that touch that though are deprecated for precisely this reason.